### PR TITLE
Feat: Add basic document management

### LINKED ISSUE #1: [Server] Basic Document Management
## Motivation

Alexandria needs server-side document management to store, retrieve, and organize user-uploaded documents. This is the foundation for all AI-powered features (summarization, entity extraction, Q&A).

## Proposed change

Implement Document domain model and service layer based on the analysis object model in `docs/diagrams/`. This includes:

- **Document** entity with owner relationship, file metadata, and raw text content
- **Tag** entity with TagSource enum (AUTO, USER) for document categorization
- **Summary** entity (1:1 with Document) for AI-generated summaries
- **ExtractedEntity** entity for AI-extracted entities (PERSON, DATE, TOPIC, ORGANIZATION)
- **FileContent** entity for binary file storage
- **SearchQuery** and **QAInteraction** entities for history tracking
- **KnowledgeBaseService** orchestrating document operations and GenAI integration

### Key Design Decisions

- Documents belong to a User (owner, via OIDC subject)
- Document fields: id, fileName, filePath, fileType, fileSize, rawTextContent, createdAt
- GenAI integration is graceful: if GenAI unavailable, document ops succeed without AI features
- Search queries and Q&A interactions are tracked for user history

### Core Entities
- [ ] #51 - Create Document entity with JPA annotations
- [ ] #52 - Create Tag entity and TagSource enum
- [ ] #100 - Create Summary and ExtractedEntity entities
- [ ] #101 - Create SearchQuery and QAInteraction entities

### Repository & Service Layer
- [ ] #53 - Create DocumentRepository interface
- [ ] #54 - Implement DocumentService with CRUD operations
- [ ] #98 - Create KnowledgeBaseService for document orchestration

### Testing
- [ ] #55 - Add unit tests for DocumentService
- [ ] #99 - Add unit tests for KnowledgeBaseService
- [ ] #56 - Add integration tests for document persistence

## Affects

- [x] Server (Spring)
- [x] GenAI (stub endpoints)
- [x] API spec
- [ ] CI / CD
- [x] Docs

## Definition of Done

- [ ] Document entity persists to database with all fields (id, fileName, filePath, fileType, fileSize, rawTextContent, createdAt)
- [ ] Tag entity with TagSource enum (AUTO, USER) persists correctly
- [ ] Document-Tag many-to-many relationship works
- [ ] Summary and ExtractedEntity entities persist correctly (1:1 and 1:many with Document)
- [ ] SearchQuery and QAInteraction entities persist correctly
- [ ] KnowledgeBaseService orchestrates document creation with GenAI calls
- [ ] DocumentService unit tests pass
- [ ] KnowledgeBaseService unit tests pass
- [ ] DocumentRepository integration tests pass
- [ ] GenAI stub endpoints exist for local testing (/genai/summarize, /genai/extract, /genai/ask)

## Branch

Implementation: `feat/basic-document-management`

### LINKED ISSUE #51: [Server] Create Document entity with JPA annotations
Part of #1.

Create the Document domain entity based on the analysis object model (`docs/diagrams/analysis_object_model.json`).

**Acceptance criteria:**

- [ ] Document entity compiles and is picked up by Spring Data
- [ ] All fields from analysis object model are present: id, fileName, filePath, fileType, fileSize, rawTextContent, createdAt
- [ ] Owner relationship to User entity (ManyToOne)
- [ ] Relationships to Summary (1:1), ExtractedEntity (1:many), Tag (many:many), FileContent (1:1)
- [ ] Can persist and retrieve documents via repository in a test

**Tasks:**

- [ ] Create `Document` entity class in `com.alexandria.app.document`
- [ ] Add JPA annotations (`@Entity`, `@Id`, `@GeneratedValue`, `@Column`, `@ManyToOne`, etc.)
- [ ] Add appropriate types (UUID for id, Instant for createdAt, etc.)
- [ ] Set up `@PrePersist` hook for createdAt timestamp
- [ ] Add getters/setters and `addTag()`/`removeTag()` helper methods

### LINKED ISSUE #52: [Server] Create Tag entity and TagSource enum
Part of #1.

Create the Tag entity and TagSource enum based on the analysis object model (`docs/diagrams/analysis_object_model.json`).

**Acceptance criteria:**

- [ ] Tag entity compiles and is picked up by Spring Data
- [ ] TagSource enum has values: AUTO, USER
- [ ] Tag has fields: id, label, source
- [ ] Many-to-many relationship with Document (Document is owning side)
- [ ] Can persist and retrieve tags via repository

**Tasks:**

- [ ] Create `TagSource` enum with AUTO and USER values
- [ ] Create `Tag` entity class with JPA annotations
- [ ] Set up the many-to-many relationship with Document (`@ManyToMany(mappedBy = "tags")`)
- [ ] Create `TagRepository` interface extending `JpaRepository<Tag, UUID>`
- [ ] Add `findByLabel()` query method

### LINKED ISSUE #53: [Server] Create DocumentRepository interface
Part of #1.

Create the Spring Data JPA repository for Document entity.

**Acceptance criteria:**

- [ ] DocumentRepository interface exists and extends JpaRepository
- [ ] Basic CRUD operations work (inherited from JpaRepository)
- [ ] Custom query methods for owner-based filtering
- [ ] Repository is autowirable in services

**Tasks:**

- [ ] Create `DocumentRepository` interface extending `JpaRepository<Document, UUID>`
- [ ] Add `findByOwnerId(UUID ownerId)` query method
- [ ] Add `findByOwnerIdAndFileNameContainingIgnoreCase(UUID ownerId, String fileName)` for search
- [ ] Add `existsByIdAndOwnerId(UUID id, UUID ownerId)` for ownership verification

### LINKED ISSUE #54: [Server] Implement DocumentService with CRUD operations
Part of #1.

Implement the service layer for document management operations.

**Acceptance criteria:**

- [ ] DocumentService supports save, findById, findByOwnerId, delete operations
- [ ] Service properly delegates to repository
- [ ] `DocumentNotFoundException` thrown for not-found cases
- [ ] Ownership verification on delete (only owner can delete)
- [ ] Service is injectable via constructor injection

**Tasks:**

- [ ] Create `DocumentService` class with `@Service` annotation
- [ ] Inject `DocumentRepository` via constructor
- [ ] Implement `save(Document)`, `findById(UUID)`, `findByOwnerId(UUID)`, `delete(UUID, UUID)`
- [ ] Create `DocumentNotFoundException` for missing documents

### LINKED ISSUE #55: [Server] Add unit tests for DocumentService
Part of #1.

Add unit tests for the DocumentService layer using mocked repository.

**Acceptance criteria:**

- [ ] Unit tests cover save, findById, findByOwnerId, delete operations
- [ ] Tests use mocked DocumentRepository (no real DB)
- [ ] Edge cases tested (not found, wrong owner)
- [ ] Tests run in CI
- [ ] Diverse file types used in test data (not just PDF)

**Tasks:**

- [ ] Create `DocumentServiceTest` class with `@ExtendWith(MockitoExtension.class)`
- [ ] Mock `DocumentRepository` using `@Mock`
- [ ] Test `save()` - verify repository save called
- [ ] Test `findById()` - happy path and not-found case
- [ ] Test `findByOwnerId()` - returns documents and empty list
- [ ] Test `delete()` - happy path, not-found, and wrong-owner cases

### LINKED ISSUE #56: [Server] Add integration tests for document persistence
Part of #1.

Add integration tests that verify document persistence against a real database (H2 in-memory).

**Acceptance criteria:**

- [ ] Integration tests verify actual database operations
- [ ] Document entity persists and retrieves correctly
- [ ] Owner-based queries work correctly
- [ ] Search by filename (case-insensitive) works
- [ ] Tests run in CI
- [ ] Diverse file types used in test data

**Tasks:**

- [ ] Create `DocumentRepositoryIntegrationTest` class
- [ ] Use `@DataJpaTest` with `@ActiveProfiles("test")`
- [ ] Test save and findById
- [ ] Test findByOwnerId (own documents and other user's documents)
- [ ] Test findByOwnerIdAndFileNameContainingIgnoreCase
- [ ] Test existsByIdAndOwnerId (true, false for wrong owner, false for nonexistent)
- [ ] Test delete operation
- [ ] Test rawTextContent storage

### LINKED ISSUE #98: [Server] Create KnowledgeBaseService for document orchestration
Part of #1.

## Motivation

The KnowledgeBaseService orchestrates document operations and integrates with the GenAI service for AI-powered features (summarization, entity extraction, Q&A).

## Proposed change

Create a service that:
- Creates documents with automatic GenAI processing (summary + entity extraction)
- Provides document CRUD with ownership verification
- Handles search with query history tracking
- Manages Q&A interactions with GenAI
- Manages document tags (add/remove)

### API

| Method | Description |
|--------|-------------|
| `createDocument(owner, fileName, filePath, fileType, fileSize, textContent)` | Create document, call GenAI for summary/entities |
| `getDocuments(ownerId)` | List user's documents |
| `getDocument(id, ownerId)` | Get single document with ownership check |
| `deleteDocument(id, ownerId)` | Delete with ownership check |
| `search(user, queryText)` | Search documents, log query history |
| `ask(user, question)` | Q&A via GenAI, log interaction |
| `addTag(documentId, ownerId, label, source)` | Add tag to document |
| `removeTag(documentId, ownerId, tagId)` | Remove tag from document |
| `getQAHistory(userId)` | Get user's Q&A history |
| `getSearchHistory(userId)` | Get user's search history |

### GenAI Integration

- Calls `/genai/summarize` and `/genai/extract` on document creation
- Calls `/genai/ask` for Q&A
- Graceful degradation: if GenAI fails, document ops succeed without AI features

**Acceptance criteria:**

- [ ] KnowledgeBaseService implements all methods above
- [ ] GenAI calls are made via GenAiClient
- [ ] Failures in GenAI don't break document creation
- [ ] Ownership verified on all document operations
- [ ] Search and Q&A history persisted

### LINKED ISSUE #99: [Server] Add unit tests for KnowledgeBaseService
Part of #1.

## Motivation

KnowledgeBaseService is the core orchestration layer for document management and GenAI integration. Comprehensive unit tests ensure correct behavior and graceful degradation.

## Proposed change

Add unit tests covering all KnowledgeBaseService methods with mocked dependencies.

**Acceptance criteria:**

- [ ] Unit tests cover all public methods
- [ ] Tests use mocked repositories and GenAiClient
- [ ] GenAI failure scenarios tested (service continues without AI features)
- [ ] Ownership verification tested
- [ ] Diverse file types used in test data
- [ ] Tests run in CI

**Test cases:**

- `createDocument` - without content (no GenAI calls)
- `createDocument` - with content (GenAI called)
- `createDocument` - GenAI fails (document created anyway)
- `getDocuments` - returns user's documents
- `getDocument` - happy path
- `getDocument` - wrong owner throws exception
- `getDocument` - not found throws exception
- `deleteDocument` - happy path
- `deleteDocument` - not found throws exception
- `search` - returns matching documents, logs query
- `ask` - calls GenAI, saves interaction
- `addTag` - creates new tag
- `addTag` - uses existing tag
- `removeTag` - removes tag from document
- `getQAHistory` / `getSearchHistory` - returns user history

### LINKED ISSUE #100: [Server] Create Summary and ExtractedEntity entities
Part of #1.

## Motivation

AI-generated summaries and extracted entities need to be persisted alongside documents for later retrieval and display.

## Proposed change

Create JPA entities for Summary and ExtractedEntity based on the analysis object model.

### Summary Entity

| Field | Type | Description |
|-------|------|-------------|
| id | UUID | Primary key |
| document | Document | 1:1 relationship (owning side) |
| content | String (TEXT) | The summary text |
| generatedAt | Instant | When summary was generated |
| modelUsed | String | AI model identifier |

### ExtractedEntity Entity

| Field | Type | Description |
|-------|------|-------------|
| id | UUID | Primary key |
| document | Document | ManyToOne relationship |
| name | String | Entity name (e.g., "John Smith") |
| type | EntityType | PERSON, DATE, TOPIC, ORGANIZATION |
| confidence | Double | Confidence score (0.0-1.0) |

### EntityType Enum

```java
public enum EntityType {
    PERSON, DATE, TOPIC, ORGANIZATION
}
```

**Acceptance criteria:**

- [ ] Summary entity with 1:1 Document relationship
- [ ] ExtractedEntity with ManyToOne Document relationship
- [ ] EntityType enum with four values
- [ ] Repositories for both entities
- [ ] Cascade delete when Document is deleted

### LINKED ISSUE #101: [Server] Create SearchQuery and QAInteraction entities
Part of #1.

## Motivation

Track user search queries and Q&A interactions for history display and analytics.

## Proposed change

Create JPA entities to persist search and Q&A history.

### SearchQuery Entity

| Field | Type | Description |
|-------|------|-------------|
| id | UUID | Primary key |
| user | User | ManyToOne relationship |
| queryText | String | The search query |
| timestamp | Instant | When query was made |
| resultCount | Integer | Number of results returned |

### QAInteraction Entity

| Field | Type | Description |
|-------|------|-------------|
| id | UUID | Primary key |
| user | User | ManyToOne relationship |
| question | String (TEXT) | User's question |
| answer | String (TEXT) | AI-generated answer |
| sourceDocumentIds | List<UUID> | Documents used for answer |
| timestamp | Instant | When interaction occurred |
| modelUsed | String | AI model identifier |

**Acceptance criteria:**

- [ ] SearchQuery entity with User relationship
- [ ] QAInteraction entity with User relationship
- [ ] sourceDocumentIds stored via `@ElementCollection`
- [ ] Repositories with `findByUserIdOrderByTimestampDesc` query
- [ ] Timestamps set automatically

### PR DESCRIPTION
**Pull request by ladu-tu**
## What does this PR do?

This PR closes #1 and its subissues regarding the document management for the spring service.

Detailled summary:
- Added JPA Document entity and repository/service. Closes #51, closes #53, closes #54.
- Added KnowledgeBase entity. Closes #98.
- Added Tag entity and corresponding TagSource enum. Closes #52 
- Added GenAI related entities. Closes #100, closes #101.
- Added unit and integration tests. Closes #55, closes #99, closes #56.

## Type of change

- [ ] Bug fix
- [X] New feature
- [ ] Refactor / cleanup
- [X] Docs
- [x] CI / infra

## Definition of Done

- [x] CI is green
- [x] If the API changed: `api/openapi.yaml` is updated and lint passes
- [X] If a service was added or restructured: docker-compose still comes up cleanly (`docker compose up`)
- [X] Tests added or updated for the changed behaviour
- [X] Docs updated where it matters (README, ADRs, comments)

## Notes for the reviewer

Note that I also added some stubs to the GenAI service. The implementation of those should be done according to #94.

### CODE REVIEW

File: services/spring/app/src/main/java/com/alexandria/app/document/Document.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+    public Summary getSummary() {
+        return summary;
+    }
+
+    public void setSummary(Summary summary) {
+        this.summary = summary;
+    }
+
+    public List<ExtractedEntity> getExtractedEntities() {
```
[the remark this answers is not in the document]
Comment (ladu-tu) [reply services/spring/app/src/main/java/com/alexandria/app/document/Document.java:150]: Suggestion seems sensible to me. I applied it.

File: services/spring/app/src/main/java/com/alexandria/app/document/Tag.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    public TagSource getSource() {
+        return source;
+    }
+
+    public void setSource(TagSource source) {
+        this.source = source;
+    }
+
+    public Set<Document> getDocuments() {
```
[the remark this answers is not in the document]
Comment (ladu-tu) [reply services/spring/app/src/main/java/com/alexandria/app/document/Tag.java:54]: Same as above. I applied the suggestion.

**Review by DoPri — APPROVED** [review]
Ok

**Comment by DoPri** [comment]
Please remember to run the pre-commit checks. I'm seeing many warnings for Spring-related stuff.

**Comment by ladu-tu** [comment]
> Please remember to run the pre-commit checks. I'm seeing many warnings for Spring-related stuff.

@DoPri Which warnings do you mean? The line endings check? I've rerun it on all files now.

**Comment by DoPri** [comment]
Yes. And thanks!
